packages/report-flaky-tests/lib/markdown.ts

```diff
@@ -0,0 +1,201 @@
+
+## Test path
+\`${ testPath }\`
+
+## Errors
+${ TEST_RESULTS_LIST.open }
+${ formattedTestResults }
+${ TEST_RESULTS_LIST.close }
+`;
+}
+
+function formatTestErrorMessage( flakyTestResult: FlakyTestResult ) {
+	switch ( flakyTestResult.runner ) {
+		case '@playwright/test': {
+			// Could do a slightly better formatting than this.
+			return stripAnsi(
+				flakyTestResult.results
+					.map( ( result ) => result!.error!.stack )
+					.join( '\n' )
+			);
+		}
+		case 'jest-circus':
+		default: {
+			return stripAnsi(
+				formatResultsErrors(
+					flakyTestResult.results,
+					{
+						rootDir: path.join(
+							process.cwd(),
+							'packages/e2e-tests'
+						),
+						// This is useless just to make typescript happy.
+						testMatch: [],
+					},
+					{ noStackTrace: false },
+					flakyTestResult.path
+				)!
+			);
+		}
+	}
+}
+
+function formatTestResults( {
+	date,
+	failedTimes,
+	headBranch,
+	runURL,
+	errorMessage,
+}: ParsedTestResult ) {
+	const dateString = date.toISOString();
+
+	// It will look something like this without formatting:
+	// [2021-08-31T16:15:19.875Z] Test passed after 2 failed attempts on trunk.
+	const log = `<time datetime="${ dateString }"><code>[${ dateString }]</code></time> Test passed after ${ failedTimes } failed ${
+		failedTimes === 1 ? 'attempt' : 'attempts'
+	} on <a href="${ runURL }"><code>${ headBranch }</code></a>.`;
+
+	if ( ! errorMessage ) {
+		return `${ TEST_RESULT.open }${ log }${ TEST_RESULT.close }`;
+	}
+
+	return `${ TEST_RESULT.open }<details>
+<summary>
+	${ log }
+</summary>
+
+\`\`\`
+${ errorMessage }
+\`\`\`
+</details>${ TEST_RESULT.close }`;
+}
+
+function parseFormattedTestResults(
+	formattedTestResults: string
+): {
+	date: Date;
+	failedTimes: number;
+	runURL: string;
+	headBranch: string;
+	errorMessage?: string;
+} {
+	const matches = formattedTestResults
+		// Unify line breaks.
+		.replace( /\r\n/g, '\n' )
+		.match( TEST_FULL_REGEX );
+
+	if ( ! matches ) {
+		throw new Error( `Unable to parse the test results:
+${ formattedTestResults }` );
+	}
+
+	const {
+		date,
+		failedTimes,
+		runURL,
+		headBranch,
+		errorMessage,
+	} = matches.groups!;
+
+	return {
+		date: new Date( date ),
+		failedTimes: parseInt( failedTimes, 10 ),
+		runURL,
+		headBranch,
+		errorMessage,
+	};
+}
+
+function parseIssueBody( body: string ) {
+	const meta = metaData.get( body );
+
+	if ( ! meta ) {
+		throw new Error( 'Unable to parse meta data from issue.' );
+	}
+
+	const testResults = Array.from(
+		body.matchAll(
+			new RegExp(
+				`^${ TEST_RESULT.open }[\\s\\S]+?${ TEST_RESULT.close }$`,
+				'gm'
+			)
+		)
+	)
+		.map( ( [ match ] ) => match )
+		.map( ( testResult ) => {
+			try {
+				return parseFormattedTestResults( testResult );
+			} catch ( error ) {
+				core.error( error as Error );
+				return undefined;
+			}
+		} )
+		.filter(
+			(
+				testResult
+			): testResult is ReturnType< typeof parseFormattedTestResults > =>
+				testResult !== undefined
+		);
```

Comment:
I think this would read simpler as a `for ... of` as then you wouldn't need so many `map`s or the `is` type assertion.

Reply:
I think `map`s are simpler though 😅, and `for...of` doesn't play nice with TypeScript I believe.

Reply:
OK, lets' call it developer preference!

But at least get rid of the `ReturnType< typeof parseFormattedTestResults >`. If you add a new `type` declaration—e.g. `TestResult`—for the return type of `parseFormattedTestResults` then you can write `TestResult` instead of `ReturnType< typeof parseFormattedTestResults >`. This is easier to read and offers more value because we're asserting something useful instead of writing the same code twice: once in the type declaration and once in the body.

Reply:
True! Agree about that 👍